src/serverenvironment.cpp

```diff
@@ -1922,16 +1927,14 @@ u16 ServerEnvironment::addActiveObjectRaw(std::unique_ptr<ServerActiveObject> ob
 			errorstream << "ServerEnvironment::addActiveObjectRaw(): "
 				<< "could not emerge block " << p << " for storing id="
 				<< object->getId() << " statically" << std::endl;
+			// clean in case of error
+			object->removingFromEnvironment();
+			m_script->removeObjectReference(object);
 			m_ao_manager.removeObject(object->getId());
 			return 0;
 		}
 	}
 
-	// Register reference in scripting api (must be done before post-init)
-	m_script->addObjectReference(object);
-	// Post-initialize object
-	object->addedToEnvironment(dtime_s);
```

Comment:
why?

Reply:
Because
https://github.com/minetest/minetest/blob/ca0a57d8eafc8079e23d8ad053d72fb81c5c6235/src/staticobject.cpp#L24-L29
`StaticObject` constructor calls `get_staticdata` from the lua entity -> https://github.com/minetest/minetest/blob/ca0a57d8eafc8079e23d8ad053d72fb81c5c6235/src/server/luaentity_sao.cpp#L286

So it should be called after `on_activate` and `on_activate` is called from  https://github.com/minetest/minetest/blob/ca0a57d8eafc8079e23d8ad053d72fb81c5c6235/src/server/luaentity_sao.cpp#L99

So moving these calls after https://github.com/minetest/minetest/blob/ca0a57d8eafc8079e23d8ad053d72fb81c5c6235/src/serverenvironment.cpp#L1913 can result in saving uninitialized data and also modder expect in `get_staticdata` call, that `on_activate` has been processed, and some additional fields in `luaentity` has been set. So there is a bigger chance of Lua error in `get_staticdata` call if `get_staticdata` will be called before `on_activate`.

Reply:
`LuaEntitySAO::getStaticData` only interacts with the script api if `m_registered` is true.
`m_registered` is only set to true in `LuaEntitySAO::addedToEnvironment`

Reply:
I missed this. Changed back.

Checked with entity callbacks:callback, `get_staticdata` is still called before saving.